File: agenkit-zig/src/middleware/caching.zig
/// Caching middleware with LRU eviction and TTL expiration
///
/// The caching middleware reduces latency and cost by caching agent responses.
/// It implements:
/// - LRU (Least Recently Used) eviction when cache is full
/// - TTL (Time To Live) based expiration with automatic cleanup
/// - Cache invalidation (specific entries or entire cache)
/// - Thread-safe operations with std.Thread.RwLock
/// - Comprehensive metrics (hits, misses, hit rate, evictions)
///
/// Example:
/// ```zig
/// const config = CachingConfig{
///     .max_cache_size = 1000,
///     .default_ttl_ms = 300000,  // 5 minutes
/// };
///
/// var caching_agent = try CachingDecorator.init(allocator, base_agent, config);
/// defer caching_agent.deinit();
///
/// const result = try caching_agent.agent().process(message);
/// const metrics = caching_agent.metrics();
/// ```
const std = @import("std");
const Agent = @import("../agent.zig").Agent;
const AgentError = @import("../agent.zig").AgentError;
const Result = @import("../agent.zig").Result;
const Message = @import("../message.zig").Message;
const IntrospectionResult = @import("../introspection.zig").IntrospectionResult;
const createDefaultIntrospectionResult = @import("../introspection.zig").createDefaultIntrospectionResult;
const Allocator = std.mem.Allocator;

/// Configuration for caching behavior
pub const CachingConfig = struct {
    /// Maximum number of entries in cache
    /// Default: 1000
    max_cache_size: u32 = 1000,

    /// Time-to-live for cache entries (milliseconds)
    /// Default: 300000ms (5 minutes)
    default_ttl_ms: u64 = 300000,

    /// Validate configuration
    pub fn validate(self: CachingConfig) !void {
        if (self.max_cache_size < 1) {
            return error.InvalidConfig;
        }
        if (self.default_ttl_ms == 0) {
            return error.InvalidConfig;
        }
    }
};

/// Cache entry with expiration and LRU tracking
const CacheEntry = struct {
    response: Message,
    expires_at_ms: i64,
    created_at_ms: i64,
    last_access_ms: i64, // For LRU tracking

    fn isExpired(self: *const CacheEntry) bool {
        return std.time.milliTimestamp() >= self.expires_at_ms;
    }
};

/// Metrics tracked by caching middleware
pub const CachingMetrics = struct {
    /// Total number of requests
    total_requests: u64 = 0,

    /// Number of cache hits
    cache_hits: u64 = 0,

    /// Number of cache misses
    cache_misses: u64 = 0,

    /// Number of evictions (LRU + expired)
    evictions: u64 = 0,

    /// Number of invalidations
    invalidations: u64 = 0,

    /// Current cache size
    current_size: u64 = 0,

    /// Cache hit rate (percentage)
    pub fn hitRate(self: *const CachingMetrics) ?f64 {
        if (self.total_requests == 0) return null;
        return (@as(f64, @floatFromInt(self.cache_hits)) / @as(f64, @floatFromInt(self.total_requests))) * 100.0;
    }

    /// Cache miss rate (percentage)
    pub fn missRate(self: *const CachingMetrics) ?f64 {
        if (self.total_requests == 0) return null;
        return (@as(f64, @floatFromInt(self.cache_misses)) / @as(f64, @floatFromInt(self.total_requests))) * 100.0;
    }

    /// Create a snapshot of metrics
    pub fn snapshot(self: *const CachingMetrics) CachingMetrics {
        return CachingMetrics{
            .total_requests = self.total_requests,
            .cache_hits = self.cache_hits,
            .cache_misses = self.cache_misses,
            .evictions = self.evictions,
            .invalidations = self.invalidations,
            .current_size = self.current_size,
        };
    }
};

/// Caching decorator - wraps an agent with caching
pub const CachingDecorator = struct {
    allocator: Allocator,
    inner_agent: Agent,
    config: CachingConfig,
    metrics_data: CachingMetrics,
    cache: std.StringHashMap(CacheEntry),
    rwlock: std.Thread.RwLock,
    cleanup_counter: u64,

    pub fn init(allocator: Allocator, inner_agent: Agent, config: CachingConfig) !*CachingDecorator {
        // Validate configuration
        try config.validate();

        const self = try allocator.create(CachingDecorator);
        self.* = CachingDecorator{
            .allocator = allocator,
            .inner_agent = inner_agent,
            .config = config,
            .metrics_data = CachingMetrics{},
            .cache = std.StringHashMap(CacheEntry).init(allocator),
            .rwlock = std.Thread.RwLock{},
            .cleanup_counter = 0,
        };
        return self;
    }

    pub fn deinit(self: *CachingDecorator) void {
        // Clean up cache entries
        var iter = self.cache.iterator();
        while (iter.next()) |entry| {
            // Free the key (owned by hashmap)
            self.allocator.free(entry.key_ptr.*);
            // Response message is owned by cache, deinit it
            entry.value_ptr.response.deinit();
        }
        self.cache.deinit();
    }

    pub fn agent(self: *CachingDecorator) Agent {
        return Agent{
            .ptr = self,
            .vtable = &.{
                .name = nameImpl,
                .capabilities = capabilitiesImpl,
                .process = processImpl,
                .introspect = introspectImpl,
                .deinit = deinitImpl,
            },
        };
    }

    /// Get current metrics (thread-safe)
    pub fn metrics(self: *CachingDecorator) CachingMetrics {
        self.rwlock.lockShared();
        defer self.rwlock.unlockShared();
        return self.metrics_data.snapshot();
    }

    /// Generate cache key from message
    fn generateCacheKey(self: *CachingDecorator, message: Message) ![]const u8 {
        // Hash role + content + metadata
        var hasher = std.crypto.hash.sha2.Sha256.init(.{});

        // Add role
        const role_str = @tagName(message.role);
        hasher.update(role_str);

        // Add content
        const content_str = message.content();
        hasher.update(content_str);

        // Add metadata (simplified: just hash keys and values)
        var meta_iter = message.metadata.iterator();
        while (meta_iter.next()) |entry| {
            hasher.update(entry.key_ptr.*);
            hasher.update(entry.value_ptr.*);
        }

        var hash_bytes: [32]u8 = undefined;
        hasher.final(&hash_bytes);

        // Convert to hex string
        const key = try self.allocator.alloc(u8, 64); // 32 bytes * 2 hex chars
        _ = std.fmt.bufPrint(key, "{s}", .{std.fmt.fmtSliceHexLower(&hash_bytes)}) catch unreachable;
        return key;
    }

    /// Cleanup expired entries
    fn cleanupExpired(self: *CachingDecorator) void {
        var keys_to_remove = std.ArrayList([]const u8).init(self.allocator);
        defer keys_to_remove.deinit();

        // Find expired entries
        var iter = self.cache.iterator();
        while (iter.next()) |entry| {
            if (entry.value_ptr.isExpired()) {
                keys_to_remove.append(entry.key_ptr.*) catch continue;
            }
        }

        // Remove expired entries
        for (keys_to_remove.items) |key| {
            if (self.cache.fetchRemove(key)) |kv| {
                self.allocator.free(kv.key);
                kv.value.response.deinit();
                self.metrics_data.evictions += 1;
            }
        }

        if (keys_to_remove.items.len > 0) {
            self.metrics_data.current_size = self.cache.count();
        }
    }

    /// Evict LRU entry when cache is full
    fn evictLRU(self: *CachingDecorator) void {
        if (self.cache.count() >= self.config.max_cache_size) {
            // Find oldest entry (smallest last_access_ms)
            var oldest_key: ?[]const u8 = null;
            var oldest_time: i64 = std.math.maxInt(i64);

            var iter = self.cache.iterator();
            while (iter.next()) |entry| {
                if (entry.value_ptr.last_access_ms < oldest_time) {
                    oldest_time = entry.value_ptr.last_access_ms;
                    oldest_key = entry.key_ptr.*;
                }
            }

            // Remove oldest entry
            if (oldest_key) |key| {
                if (self.cache.fetchRemove(key)) |kv| {
                    self.allocator.free(kv.key);
                    kv.value.response.deinit();
                    self.metrics_data.evictions += 1;
                    self.metrics_data.current_size = self.cache.count();
                }
            }
        }
    }

    /// Invalidate cache (entire cache or specific message)
    pub fn invalidate(self: *CachingDecorator, message: ?Message) !void {
        self.rwlock.lock();
        defer self.rwlock.unlock();

        if (message) |msg| {
            // Invalidate specific entry
            const key = try self.generateCacheKey(msg);
            defer self.allocator.free(key);

            if (self.cache.fetchRemove(key)) |kv| {
                self.allocator.free(kv.key);
                kv.value.response.deinit();
                self.metrics_data.invalidations += 1;
                self.metrics_data.current_size = self.cache.count();
            }
        } else {
            // Invalidate entire cache
            var count: u64 = 0;
            var iter = self.cache.iterator();
            while (iter.next()) |entry| {
                self.allocator.free(entry.key_ptr.*);
                entry.value_ptr.response.deinit();
                count += 1;
            }
            self.cache.clearRetainingCapacity();
            self.metrics_data.invalidations += count;
            self.metrics_data.current_size = 0;
        }
    }

    fn nameImpl(ptr: *anyopaque) []const u8 {
        const self: *CachingDecorator = @ptrCast(@alignCast(ptr));
        return self.inner_agent.name();
    }

    fn capabilitiesImpl(ptr: *anyopaque, allocator: Allocator) Allocator.Error![]const []const u8 {
        const self: *CachingDecorator = @ptrCast(@alignCast(ptr));
        return self.inner_agent.capabilities(allocator);
    }

    fn processImpl(ptr: *anyopaque, message: Message) AgentError!Result {
        const self: *CachingDecorator = @ptrCast(@alignCast(ptr));

        // Generate cache key (no lock needed)
        const cache_key = self.generateCacheKey(message) catch {
            // If key generation fails, bypass cache
            return self.inner_agent.process(message);
        };
        defer self.allocator.free(cache_key);

        // Check cache with read lock
        self.rwlock.lockShared();
        if (self.cache.get(cache_key)) |entry| {
            if (!entry.isExpired()) {
                // Cache hit
                const response = entry.response;
                self.rwlock.unlockShared();

                // Update metrics (outside lock)
                self.rwlock.lock();
                self.metrics_data.total_requests += 1;
                self.metrics_data.cache_hits += 1;
                self.rwlock.unlock();

                return Result{ .ok = response };
            }
        }
        self.rwlock.unlockShared();

        // Cache miss - update metrics
        self.rwlock.lock();
        self.metrics_data.total_requests += 1;
        self.metrics_data.cache_misses += 1;
        self.cleanup_counter += 1;
        const do_cleanup = self.cleanup_counter % 100 == 0;
        self.rwlock.unlock();

        // Process message (outside lock)
        const result = try self.inner_agent.process(message);
        const response = try result.unwrap();

        // Cache response (write lock)
        self.rwlock.lock();
        defer self.rwlock.unlock();

        // Periodic cleanup
        if (do_cleanup) {
            self.cleanupExpired();
        }

        // Evict LRU if needed
        self.evictLRU();

        // Add to cache
        const now_ms = std.time.milliTimestamp();
        const key_owned = try self.allocator.dupe(u8, cache_key);
        const entry = CacheEntry{
            .response = response,
            .expires_at_ms = now_ms + @as(i64, @intCast(self.config.default_ttl_ms)),
            .created_at_ms = now_ms,
            .last_access_ms = now_ms,
        };
        try self.cache.put(key_owned, entry);
        self.metrics_data.current_size = self.cache.count();

        return Result{ .ok = response };
    }

    fn introspectImpl(ptr: *anyopaque, allocator: Allocator) Allocator.Error!IntrospectionResult {
        const self: *CachingDecorator = @ptrCast(@alignCast(ptr));

        // Get introspection from inner agent
        const inner_result = try self.inner_agent.introspect(allocator);

        // Add caching metrics to metadata
        const metrics_snapshot = self.metrics();

        var metadata = std.StringHashMap([]const u8).init(allocator);
        errdefer metadata.deinit();

        // Add metrics as metadata
        try metadata.put("total_requests", try std.fmt.allocPrint(allocator, "{d}", .{metrics_snapshot.total_requests}));
        try metadata.put("cache_hits", try std.fmt.allocPrint(allocator, "{d}", .{metrics_snapshot.cache_hits}));
        try metadata.put("cache_misses", try std.fmt.allocPrint(allocator, "{d}", .{metrics_snapshot.cache_misses}));
        try metadata.put("evictions", try std.fmt.allocPrint(allocator, "{d}", .{metrics_snapshot.evictions}));
        try metadata.put("invalidations", try std.fmt.allocPrint(allocator, "{d}", .{metrics_snapshot.invalidations}));
        try metadata.put("current_size", try std.fmt.allocPrint(allocator, "{d}", .{metrics_snapshot.current_size}));

        if (metrics_snapshot.hitRate()) |rate| {
            try metadata.put("hit_rate", try std.fmt.allocPrint(allocator, "{d:.2}%", .{rate}));
        }
        if (metrics_snapshot.missRate()) |rate| {
            try metadata.put("miss_rate", try std.fmt.allocPrint(allocator, "{d:.2}%", .{rate}));
        }

        // Add configuration
        try metadata.put("max_cache_size", try std.fmt.allocPrint(allocator, "{d}", .{self.config.max_cache_size}));
        try metadata.put("default_ttl_ms", try std.fmt.allocPrint(allocator, "{d}", .{self.config.default_ttl_ms}));

        // Merge with inner metadata
        var inner_iter = inner_result.metadata.iterator();
        while (inner_iter.next()) |entry| {
            if (!std.mem.startsWith(u8, entry.key_ptr.*, "cache_")) {
                try metadata.put(entry.key_ptr.*, entry.value_ptr.*);
            }
        }

        return IntrospectionResult{
            .agent_name = inner_result.agent_name,
            .capabilities = inner_result.capabilities,
            .metadata = metadata,
            .memory = inner_result.memory,
        };
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *CachingDecorator = @ptrCast(@alignCast(ptr));
        self.deinit();
        self.allocator.destroy(self);
    }
};

// Tests
const testing = std.testing;

test "CachingConfig validation" {
    var config = CachingConfig{};
    try config.validate();

    // Invalid: max_cache_size < 1
    config.max_cache_size = 0;
    try testing.expectError(error.InvalidConfig, config.validate());
    config.max_cache_size = 1000;

    // Invalid: default_ttl_ms == 0
    config.default_ttl_ms = 0;
    try testing.expectError(error.InvalidConfig, config.validate());
}

test "CachingMetrics calculations" {
    var metrics = CachingMetrics{
        .total_requests = 100,
        .cache_hits = 80,
        .cache_misses = 20,
    };

    // Hit rate
    const hit_rate = metrics.hitRate().?;
    try testing.expectApproxEqRel(@as(f64, 80.0), hit_rate, 0.01);

    // Miss rate
    const miss_rate = metrics.missRate().?;
    try testing.expectApproxEqRel(@as(f64, 20.0), miss_rate, 0.01);
}

File: agenkit-rust/src/middleware/caching.rs
//! Caching middleware with LRU eviction and TTL support.
//!
//! Caches agent responses to avoid redundant processing of identical requests,
//! reducing latency, cost, and load on downstream services.
//!
//! # Features
//!
//! - **LRU Eviction**: Least Recently Used items evicted when cache is full
//! - **TTL Support**: Entries expire after configured time-to-live
//! - **Thread-Safe**: Uses RwLock for concurrent access
//! - **Configurable**: Adjustable size and TTL
//!
//! # When to Use
//!
//! - **Expensive operations**: LLM calls, database queries, API requests
//! - **Repeated queries**: Same questions asked frequently
//! - **Cost reduction**: Avoid paying for duplicate API calls
//! - **Latency improvement**: Instant response from cache
//!
//! # When NOT to Use
//!
//! - **Dynamic content**: Responses change frequently
//! - **User-specific**: Responses vary per user/context
//! - **Side effects**: Operations that modify state
//! - **Real-time data**: Need fresh data every time
//!
//! # Cache Key
//!
//! Cache key is generated from message content (JSON serialization).
//! Messages with identical content share the same cache entry.
//!
//! # Example
//!
//! ```rust
//! use agenkit::middleware::{CachingMiddleware, CachingConfig};
//! use agenkit::core::{Agent, Message};
//! use std::time::Duration;
//!
//! # async fn example() {
//! # struct MyAgent;
//! # #[async_trait::async_trait]
//! # impl Agent for MyAgent {
//! #     fn name(&self) -> &str { "test" }
//! #     async fn process(&self, msg: Message) -> Result<Message, agenkit::core::AgentError> {
//! #         Ok(Message::with_text("assistant", "ok"))
//! #     }
//! # }
//! let agent = MyAgent;
//!
//! let config = CachingConfig::builder()
//!     .max_size(1000)
//!     .ttl(Duration::from_secs(300))  // 5 minutes
//!     .build();
//!
//! let cache_agent = CachingMiddleware::new(agent, config);
//!
//! // First call: hits agent
//! let msg = Message::with_text("user", "What is 2+2?");
//! let response1 = cache_agent.process(msg.clone()).await;
//!
//! // Second call: served from cache
//! let response2 = cache_agent.process(msg).await;
//! # }
//! ```

use crate::core::{Agent, AgentError, IntrospectionResult, Message};
use async_trait::async_trait;
use std::collections::HashMap;
use std::hash::{Hash, Hasher};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::RwLock;

/// Configuration for caching middleware.
#[derive(Debug, Clone)]
pub struct CachingConfig {
    /// Maximum number of entries in cache.
    /// Default: 1000
    pub max_size: usize,

    /// Time-to-live for cache entries.
    /// Default: 5 minutes
    pub ttl: Duration,
}

impl Default for CachingConfig {
    fn default() -> Self {
        Self {
            max_size: 1000,
            ttl: Duration::from_secs(300), // 5 minutes
        }
    }
}

impl CachingConfig {
    /// Create a new builder for CachingConfig.
    pub fn builder() -> CachingConfigBuilder {
        CachingConfigBuilder::default()
    }
}

/// Builder for CachingConfig.
#[derive(Debug, Default)]
pub struct CachingConfigBuilder {
    max_size: Option<usize>,
    ttl: Option<Duration>,
}

impl CachingConfigBuilder {
    /// Set maximum cache size.
    pub fn max_size(mut self, size: usize) -> Self {
        self.max_size = Some(size);
        self
    }

    /// Set time-to-live for cache entries.
    pub fn ttl(mut self, ttl: Duration) -> Self {
        self.ttl = Some(ttl);
        self
    }

    /// Build the CachingConfig.
    pub fn build(self) -> CachingConfig {
        let default = CachingConfig::default();
        CachingConfig {
            max_size: self.max_size.unwrap_or(default.max_size),
            ttl: self.ttl.unwrap_or(default.ttl),
        }
    }
}

/// Metrics for caching middleware.
#[derive(Debug, Clone, Default)]
pub struct CachingMetrics {
    /// Total number of requests processed.
    pub total_requests: u64,

    /// Number of cache hits.
    pub cache_hits: u64,

    /// Number of cache misses.
    pub cache_misses: u64,

    /// Number of evictions (LRU + expired).
    pub evictions: u64,

    /// Current cache size.
    pub current_size: usize,
}

impl CachingMetrics {
    /// Calculate cache hit rate (percentage).
    pub fn hit_rate(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            (self.cache_hits as f64 / self.total_requests as f64) * 100.0
        }
    }

    /// Calculate cache miss rate (percentage).
    pub fn miss_rate(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            (self.cache_misses as f64 / self.total_requests as f64) * 100.0
        }
    }
}

/// Cache entry with value and metadata.
#[derive(Debug, Clone)]
struct CacheEntry {
    value: Message,
    inserted_at: Instant,
    last_accessed: Instant,
}

impl CacheEntry {
    fn new(value: Message) -> Self {
        let now = Instant::now();
        Self {
            value,
            inserted_at: now,
            last_accessed: now,
        }
    }

    fn is_expired(&self, ttl: Duration) -> bool {
        self.inserted_at.elapsed() >= ttl
    }

    fn touch(&mut self) {
        self.last_accessed = Instant::now();
    }
}

/// Generate cache key from message.
fn cache_key(message: &Message) -> u64 {
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    // Hash the JSON representation of the message
    if let Ok(json) = serde_json::to_string(message) {
        json.hash(&mut hasher);
    }
    hasher.finish()
}

/// LRU cache implementation.
#[derive(Debug)]
struct LruCache {
    entries: HashMap<u64, CacheEntry>,
    max_size: usize,
    ttl: Duration,
    metrics: CachingMetrics,
}

impl LruCache {
    fn new(max_size: usize, ttl: Duration) -> Self {
        Self {
            entries: HashMap::new(),
            max_size,
            ttl,
            metrics: CachingMetrics::default(),
        }
    }

    fn get(&mut self, key: u64) -> Option<Message> {
        // Check if entry exists and is not expired
        let is_expired = if let Some(entry) = self.entries.get(&key) {
            entry.is_expired(self.ttl)
        } else {
            return None;
        };

        if is_expired {
            self.entries.remove(&key);
            self.metrics.current_size = self.entries.len();
            return None;
        }

        // Entry exists and is valid - touch it and return
        let result = if let Some(entry) = self.entries.get_mut(&key) {
            entry.touch();
            Some(entry.value.clone())
        } else {
            None
        };

        // Update metrics after releasing mutable borrow
        if result.is_some() {
            self.metrics.cache_hits += 1;
            self.metrics.current_size = self.entries.len();
        }

        result
    }

    fn insert(&mut self, key: u64, value: Message) {
        // Evict expired entries
        self.evict_expired();

        // Evict LRU entry if at capacity
        if self.entries.len() >= self.max_size && !self.entries.contains_key(&key) {
            self.evict_lru();
        }

        self.entries.insert(key, CacheEntry::new(value));
        self.metrics.current_size = self.entries.len();
    }

    fn evict_expired(&mut self) {
        let expired_keys: Vec<u64> = self
            .entries
            .iter()
            .filter(|(_, entry)| entry.is_expired(self.ttl))
            .map(|(k, _)| *k)
            .collect();

        let eviction_count = expired_keys.len();
        for key in expired_keys {
            self.entries.remove(&key);
        }

        if eviction_count > 0 {
            self.metrics.evictions += eviction_count as u64;
            self.metrics.current_size = self.entries.len();
        }
    }

    fn evict_lru(&mut self) {
        if let Some((&lru_key, _)) = self
            .entries
            .iter()
            .min_by_key(|(_, entry)| entry.last_accessed)
        {
            self.entries.remove(&lru_key);
            self.metrics.evictions += 1;
            self.metrics.current_size = self.entries.len();
        }
    }

    fn size(&self) -> usize {
        self.entries.len()
    }

    fn get_metrics(&self) -> CachingMetrics {
        self.metrics.clone()
    }
}

/// Caching middleware with LRU eviction and TTL support.
///
/// Caches agent responses to avoid redundant processing, improving
/// latency and reducing cost for repeated queries.
pub struct CachingMiddleware<A: Agent> {
    inner: A,
    config: CachingConfig,
    cache: Arc<RwLock<LruCache>>,
}

impl<A: Agent> CachingMiddleware<A> {
    /// Create a new caching middleware with the given agent and configuration.
    pub fn new(agent: A, config: CachingConfig) -> Self {
        Self {
            inner: agent,
            cache: Arc::new(RwLock::new(LruCache::new(config.max_size, config.ttl))),
            config,
        }
    }

    /// Create a new caching middleware with default configuration.
    pub fn with_defaults(agent: A) -> Self {
        Self::new(agent, CachingConfig::default())
    }

    /// Get current caching metrics.
    pub async fn get_metrics(&self) -> CachingMetrics {
        let cache = self.cache.read().await;
        cache.get_metrics()
    }
}

#[async_trait]
impl<A: Agent> Agent for CachingMiddleware<A> {
    fn name(&self) -> &str {
        self.inner.name()
    }

    fn capabilities(&self) -> Vec<String> {
        self.inner.capabilities()
    }

    fn introspect(&self) -> IntrospectionResult {
        let mut result = self.inner.introspect();
        result
            .metadata
            .insert("middleware".to_string(), serde_json::json!("caching"));
        result.metadata.insert(
            "caching_config".to_string(),
            serde_json::json!({
                "max_size": self.config.max_size,
                "ttl_ms": self.config.ttl.as_millis(),
            }),
        );
        result
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        let key = cache_key(&message);

        // Track total requests and try to get from cache
        {
            let mut cache = self.cache.write().await;
            cache.metrics.total_requests += 1;
            if let Some(cached) = cache.get(key) {
                return Ok(cached);
            }
        }

        // Cache miss - track and call inner agent
        {
            let mut cache = self.cache.write().await;
            cache.metrics.cache_misses += 1;
        }

        let response = self.inner.process(message).await?;

        // Store in cache
        {
            let mut cache = self.cache.write().await;
            cache.insert(key, response.clone());
        }

        Ok(response)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicU32, Ordering};

    struct CountingAgent {
        attempts: Arc<AtomicU32>,
    }

    impl CountingAgent {
        fn new() -> Self {
            Self {
                attempts: Arc::new(AtomicU32::new(0)),
            }
        }

        fn attempt_count(&self) -> u32 {
            self.attempts.load(Ordering::SeqCst)
        }
    }

    #[async_trait]
    impl Agent for CountingAgent {
        fn name(&self) -> &str {
            "counting"
        }

        async fn process(&self, message: Message) -> Result<Message, AgentError> {
            let count = self.attempts.fetch_add(1, Ordering::SeqCst);
            Ok(Message::with_text(
                "assistant",
                format!("call {}", count + 1),
            ))
        }
    }

    #[tokio::test]
    async fn test_caching_returns_cached_response() {
        let agent = CountingAgent::new();
        let config = CachingConfig::builder().max_size(10).build();

        let cache_agent = CachingMiddleware::new(agent, config);

        let msg = Message::with_text("user", "test");

        // First call - cache miss
        let result1 = cache_agent.process(msg.clone()).await.unwrap();
        assert_eq!(result1.content_as_str(), Some("call 1"));

        // Second call - cache hit
        let result2 = cache_agent.process(msg.clone()).await.unwrap();
        assert_eq!(result2.content_as_str(), Some("call 1")); // Same response

        // Agent was only called once
        assert_eq!(cache_agent.inner.attempt_count(), 1);
    }

    #[tokio::test]
    async fn test_caching_different_messages() {
        let agent = CountingAgent::new();
        let config = CachingConfig::builder().max_size(10).build();

        let cache_agent = CachingMiddleware::new(agent, config);

        let msg1 = Message::with_text("user", "test1");
        let msg2 = Message::with_text("user", "test2");

        // Different messages should not hit cache
        let result1 = cache_agent.process(msg1).await.unwrap();
        let result2 = cache_agent.process(msg2).await.unwrap();

        assert_eq!(result1.content_as_str(), Some("call 1"));
        assert_eq!(result2.content_as_str(), Some("call 2"));
        assert_eq!(cache_agent.inner.attempt_count(), 2);
    }

    #[tokio::test]
    async fn test_caching_ttl_expiration() {
        let agent = CountingAgent::new();
        let config = CachingConfig::builder()
            .ttl(Duration::from_millis(100))
            .build();

        let cache_agent = CachingMiddleware::new(agent, config);

        let msg = Message::with_text("user", "test");

        // First call
        let result1 = cache_agent.process(msg.clone()).await.unwrap();
        assert_eq!(result1.content_as_str(), Some("call 1"));

        // Wait for TTL to expire
        tokio::time::sleep(Duration::from_millis(150)).await;

        // Second call - cache expired
        let result2 = cache_agent.process(msg.clone()).await.unwrap();
        assert_eq!(result2.content_as_str(), Some("call 2"));

        assert_eq!(cache_agent.inner.attempt_count(), 2);
    }

    #[tokio::test]
    async fn test_caching_lru_eviction() {
        let agent = CountingAgent::new();
        let config = CachingConfig::builder().max_size(2).build();

        let cache_agent = CachingMiddleware::new(agent, config);

        let msg1 = Message::with_text("user", "test1");
        let msg2 = Message::with_text("user", "test2");
        let msg3 = Message::with_text("user", "test3");

        // Fill cache
        let _ = cache_agent.process(msg1.clone()).await;
        let _ = cache_agent.process(msg2.clone()).await;

        // Access msg1 to make it more recent
        let _ = cache_agent.process(msg1.clone()).await;

        // Add msg3 - should evict msg2 (LRU)
        let _ = cache_agent.process(msg3.clone()).await;

        // msg1 should still be cached
        let _ = cache_agent.process(msg1.clone()).await;

        // msg2 should not be cached
        let _ = cache_agent.process(msg2.clone()).await;

        // Verify: 1 (msg1), 2 (msg2), 3 (msg1 cached), 4 (msg3), 5 (msg1 cached), 6 (msg2 not cached)
        // Total calls: 4 (msg1, msg2, msg3, msg2 again)
        assert_eq!(cache_agent.inner.attempt_count(), 4);
    }

    #[tokio::test]
    async fn test_introspect_includes_caching_metadata() {
        let agent = CountingAgent::new();
        let config = CachingConfig::builder().max_size(100).build();

        let cache_agent = CachingMiddleware::new(agent, config);
        let result = cache_agent.introspect();

        assert_eq!(
            result.metadata.get("middleware"),
            Some(&serde_json::json!("caching"))
        );
        assert!(result.metadata.contains_key("caching_config"));
    }
}
